packages/react/__tests__/get-label-from-stack-trace.js

```diff
@@ -0,0 +1,560 @@
+  getLabelFromStackTrace
+} from '../src/get-label-from-stack-trace'
+
+describe('getFunctionNameFromStackTraceLine', () => {
+  test('V8', () => {
+    expect(getFunctionNameFromStackTraceLine('Error')).toBeUndefined()
+
+    expect(
+      getFunctionNameFromStackTraceLine(
+        '    at Object.createEmotionProps (emotion-element-1fb5ab00.browser.esm.js:143)'
+      )
+    ).toBe('createEmotionProps')
+  })
+
+  test('Safari / Firefox', () => {
+    expect(
+      getFunctionNameFromStackTraceLine(
+        'createEmotionProps@http://localhost:3000/static/js/main.chunk.js:844:40'
+      )
+    ).toBe('createEmotionProps')
+  })
+})
+
+/**
+ * # Safari stack traces
+ *
+ * The component name may not appear in the Safari stack trace because Safari
+ * implements Proper Tail Calls.
+ *
+ * Two components that seem almost identical may produce different stack traces
+ * based on whether or not a Proper Tail Call optimization was possible. As a
+ * result, you may not get the same stack traces if you try to reproduce these
+ * test cases on your own.
+ *
+ * If Safari omits the component name from the stack trace, the best thing we
+ * can do is:
+ *
+ * 1. Match the parent component's name if it is in the stack trace, or
+ * 2. Return `undefined`.
+ */
+
+// Ensure that it works for components that have numbers and $ in their name
+const expectedLabel = 'MyComponent-9'
+
+/**
+ * E.g.
+ *
+ * ```
+ * function MyComponent$9() {
+ *   return <div css={{ color: 'red' }} />
+ * }
+ * ```
+ *
+ * See `playgrounds/cra/.env` for instructions on testing with the old JSX
+ * transform.
+ *
+ * All tests after this block are for runtime=automatic, i.e. the new JSX
+ * transform introduced in React 17
+ */
+describe('typical function component - runtime=classic', () => {
+  test('Chrome', () => {
+    // Each "at" line starts with some whitespace
+    const stackTrace = `Error
+    at createEmotionProps (emotion-element-1fb5ab00.browser.esm.js:143)
+    at jsx (emotion-react.browser.esm.js:100)
+    at MyComponent$9 (App.js:22)
+    at renderWithHooks (react-dom.development.js:14803)
+    at mountIndeterminateComponent (react-dom.development.js:17482)
+    at beginWork (react-dom.development.js:18596)`
+
+    expect(getLabelFromStackTrace(stackTrace)).toBe(expectedLabel)
+  })
+
+  test('Firefox', () => {
+    const stackTrace = `createEmotionProps@http://localhost:3000/static/js/bundle.js:46440:40
+jsx@http://localhost:3000/static/js/bundle.js:46636:113
+MyComponent$9@http://localhost:3000/static/js/bundle.js:47600:72
+renderWithHooks@http://localhost:3000/static/js/bundle.js:18904:27
+mountIndeterminateComponent@http://localhost:3000/static/js/bundle.js:21583:13
+beginWork@http://localhost:3000/static/js/bundle.js:22697:16`
+
+    expect(getLabelFromStackTrace(stackTrace)).toBe(expectedLabel)
+  })
+
+  test('Safari', () => {
+    const stackTrace = `createEmotionProps@http://localhost:3000/static/js/bundle.js:46440:49
```

Comment:
is this stack trace any different - format-wise - than other stack traces in this block? it looks roughly the same as FireFox one, right?

what I'm getting at - if any of the Safari stack traces are different here we should still test them in a way that they won't result in the `undefined` label. To do that we can simply move the element out of the return position of the function, assign it to a variable and return that variable - this prevents TCO

Reply:
The Safari stack appears identical to the Firefox one apart from the component function call being omitted. So I don't think there's any need to test Safari without TCO. Let me know if you disagree and I will change it.